### PR TITLE
[Issue #630] Logging and Monitoring ADR

### LINKED ISSUE #630: [ADR]: Logging and monitoring platform for front-end and back-end
### Description

ADR to select a tool to send logging metrics such as New Relic. Consider what would be required for the first release of the Developer Preview and what tool would be set up for the long term strategy. 

### Approvers

Lucas and/or Billy 

### Options

- New Relic
- Splunk
- Cloudwatch
- Sentry
- Datadog
- Grafana

### Users
* Application developer
* HHS 
* Non-developers on high level performance

### Decision Criteria

* Interoperability: be able to connect to a visualization tool so that it can be connected with the public measurement dashboard
* API metrics: how many people are hitting the API
* Debugging capabilities
* Log searching and insights: find the issues such as slow queries that are causing issues 

### Definition of Done

- [ ] The approvers for this decision have been identified (ideally before work on the ADR starts)
- [ ] The ADR is created and stored in `documentation/decisions/adr` with status "Accepted"
- [ ] The ADR has been reviewed and approved by the approvers: Lucas and/or Billy
- [ ] The ADR satisfies requirements that are outlined in the ADR template
- [ ] Any follow-up tickets have been created to implement the logging and monitoring tool

### PR DESCRIPTION
**Pull request by coilysiren**
## Summary

Fixes #630

### CODE REVIEW

**Review by widal001 — APPROVED** [review]
This is a super helpful comparison of the options! I share your assessment that Datadog and New Relic seem like the top choices.

One thing that reading through this doc did bring to mind for me though -- would the balance of favor shift between Datadog and New Relic if we expected more of the metrics and dashboards to be handled by a data warehouse paired with a BI tool?

I see the primary goals of a monitoring tool as the following:
- Can we quickly triage, investigate, and diagnose prod incidents?
- Can we proactively monitor application load, performance, etc. to scale or adjust resources before the impact is felt by the team?

I think a lot of the product-specific monitoring goals (i.e. helping identify performance bottlenecks, publishing rates of 4xx and 5xx) could possibly be achieved with other tools (e.g. data warehouse + BI tool).

All that said I think these top two choices are still the smartest default choices and if existing ATO coverage or team comfort tips the balance for one tool or another I say we move forward with in and focus on implementaton.

File: documentation/decisions/adr/2024-03-04-logging-monitoring.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+Links:
+- [Pricing](https://www.datadoghq.com/pricing/)
+
+### New Relic
+
+New Relic is a fully featured platform with support for logging, metrics dashboards, and APM. Several members of our team have prior experience with New Relic. New Relic excels in the market, is listed on [FedRAMP](https://marketplace.fedramp.gov/products), and would be an excellent choice of platform. It has fantastic UX, for both backend and frontend applications. The only "flaw" is that, at a high level, New Relic's product offering is hard to distinguish from Datadog's. The bulk of this paragraph is copied into the Datadog description, to emphasize that point.
+
+New Relic has a free tier, and their pricing is less complex than Datadog's.
+
+New Relic was built around its APM capability, so it will have better capabilities for debugging performance issues and hunting down esoteric production bugs.
```
[the remark this answers is not in the document]
Comment (chouinar) [reply documentation/decisions/adr/2024-03-04-logging-monitoring.md:78]: `We probably don't need to capture metrics on everything (e.g. CPU utilization, load balancing, etc.)` - Why wouldn't we want those metrics? That sounds like something we'd want for alerting (eg. above 80% cpu utilization alerts).

I know from past experience, we used New Relic to fix several issues with database query optimization

File: documentation/decisions/adr/2024-03-04-logging-monitoring.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+One large advantage of Datadog is that it was built around its metrics and dashboard capabilities, and was built for API-driven applications. As such, Datadog has best-in-class dashboard functionality. This gives it a slight leg-up relative to New Relic. Datadog additionally can make dashboards public, which is a functionality that we may want to leverage to expose our API status (4XX / 5XX rate, etc) to the general public.
+
+Grants.gov used Datadog in the past and then decided to move to Cloudwatch, due to cost.
+
+- **Decision Status**: Top Choice
+- **Pros**
+  - Well-known to our team
+  - Fully featured platform (logs, metrics, APM)
+  - Strong metrics product
+  - Best for API-driven applications
```
Comment (chouinar) [inline documentation/decisions/adr/2024-03-04-logging-monitoring.md:65]: How does it do with backend applications/scripts? I know with New Relic we were able to configure the API and backend to both work pretty well.

Comment (coilysiren) [reply documentation/decisions/adr/2024-03-04-logging-monitoring.md:65]: It works great - in fact my first usage of Datadog @ my last job was as a part of a ELT script. I just attached the APM / OpenTelemetry reader to the start of the script, and it monitored it just like anything else. I assume New Relic works in exactly the same way.

File: documentation/decisions/adr/2024-03-04-logging-monitoring.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (coilysiren) [reply documentation/decisions/adr/2024-03-04-logging-monitoring.md:78]: I think there's just some miscommunication here - we will absolutely be capturing CPU / etc inside our Monitoring platform. We will also try to capture those metrics inside the BI platform for the sake of dashboards, yes.

File: documentation/decisions/adr/2024-03-04-logging-monitoring.md (hunk 2 above)
Comment (chouinar) [reply documentation/decisions/adr/2024-03-04-logging-monitoring.md:65]: I think for New Relic we had a few utilities for setting up the connection & restructuring our logs, but once that is setup, each ECS task just ended up with `@background_task("WhateverYouWantToCallYourECSTask")` on it and it just worked. Definitely has some setup, but after that it is pretty uneventful

Comment (coilysiren) [reply documentation/decisions/adr/2024-03-04-logging-monitoring.md:65]: For Datadog is looked similar, `@ddtrace.span(name="Whatever")`

There wasn't any meaningful difference between instrumenting a background task versus instrumenting a request. The majority (like 95%) of the work was sitting up the instructure layer, particularly the logging sidecars.

File: documentation/decisions/adr/2024-03-04-logging-monitoring.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (widal001) [reply documentation/decisions/adr/2024-03-04-logging-monitoring.md:78]: > I think there's just some miscommunication here - we will absolutely be capturing CPU / etc inside our Monitoring platform. We will also try to capture those metrics inside the BI platform for the sake of dashboards, yes.

Sorry! I was talking about which metrics we put onto our public measurement dashboard -- we do absolutely want to monitor CPU usage, etc. internally as part of our monitoring tool.

I think we're open to having this be on a public facing dashboard, as well, though it probably wouldn't be top priority for including it. Instead we'd probably want to focus on publish metrics on error rates, response times, availability, etc. and less around resource utilization.

**Review by lucasmbrown-usds — APPROVED** [review]
LGTM! Thanks for the thorough discussion and writeup!

**Pull request merged**